### DISCUSSION TITLE
Separate optimization for trunk and embedder

### DISCUSSION THREAD

**Discussion by lfleck** [KevinMusgrave/pytorch-metric-learning#270] [Q&A] [answered]
Hi guys, 
your examples (e.g. [TwoStreamMetricLoss](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/examples/notebooks/TwoStreamMetricLoss.ipynb)) show separate optimizers for trunk and embedder. 
Is that just for the sake of explanation how to deal with trunk and embedder in general, or is there a design decision behind? In the case of the latter, do you have any references for that? Most literature I've seen so far trains them jointly (e.g. [SimCLR](https://arxiv.org/pdf/2002.05709.pdf)). Many thanks!

**Comment by KevinMusgrave** [comment, marked as answer]
It is common in metric learning papers to use a different learning rate for the trunk and embedder. 
- The trunk is usually a model pretrained on Imagenet, so they use a lower learning rate on that part. 
- The embedder is usually just a single fully connected layer that is randomly initialized, so they use a higher learning rate on that part.

This [CVPR 2020 paper](https://arxiv.org/pdf/1912.06798.pdf) mentions using a 10 times lower learning rate on the trunk model:

![image](https://user-images.githubusercontent.com/2314891/106371002-92a69180-632d-11eb-8074-a842d2311c75.png)
